ServerCore/Pages/Teams/Play.cshtml.cs

```diff
@@ -64,41 +68,55 @@ public async Task OnGetAsync(SortOrder? sort, int teamId)
+                    visiblePuzzlesQ = visiblePuzzlesQ.OrderByDescending(pv => pv.Name);
                     break;
                 case SortOrder.GroupAscending:
-                    visiblePuzzlesQ = visiblePuzzlesQ.OrderBy(puzzleWithState => puzzleWithState.Puzzle.Group).ThenBy(puzzleWithState => puzzleWithState.Puzzle.OrderInGroup);
+                    visiblePuzzlesQ = visiblePuzzlesQ.OrderBy(pv => pv.Group).ThenBy(pv => pv.OrderInGroup);
                     break;
                 case SortOrder.GroupDescending:
-                    visiblePuzzlesQ = visiblePuzzlesQ.OrderByDescending(puzzleWithState => puzzleWithState.Puzzle.Group).ThenByDescending(puzzleWithState => puzzleWithState.Puzzle.OrderInGroup);
+                    visiblePuzzlesQ = visiblePuzzlesQ.OrderByDescending(pv => pv.Group).ThenByDescending(pv => pv.OrderInGroup);
                     break;
                 case SortOrder.SolveAscending:
-                    visiblePuzzlesQ = visiblePuzzlesQ.OrderBy(puzzleWithState => puzzleWithState.State.SolvedTime ?? DateTime.MaxValue);
+                    visiblePuzzlesQ = visiblePuzzlesQ.OrderBy(pv => pv.SolvedTime ?? DateTime.MaxValue);
                     break;
                 case SortOrder.SolveDescending:
-                    visiblePuzzlesQ = visiblePuzzlesQ.OrderByDescending(puzzleWithState => puzzleWithState.State.SolvedTime ?? DateTime.MaxValue);
+                    visiblePuzzlesQ = visiblePuzzlesQ.OrderByDescending(pv => pv.SolvedTime ?? DateTime.MaxValue);
                     break;
                 default:
                     throw new ArgumentException($"unknown sort: {sort}");
             }
 
+            PuzzleViews = await visiblePuzzlesQ.ToListAsync();
+
             Dictionary<int, ContentFile> files = await (from file in _context.ContentFiles
                                                         where file.Event == Event && file.FileType == ContentFileType.Puzzle
                                                         select file).ToDictionaryAsync(file => file.Puzzle.ID);
 
-            PuzzlesWithState = new List<PuzzleWithState>();
-            foreach(var visiblePuzzle in await visiblePuzzlesQ.ToListAsync())
+            foreach(var puzzleView in PuzzleViews)
             {
-                files.TryGetValue(visiblePuzzle.Puzzle.ID, out ContentFile content);
-                PuzzlesWithState.Add(new PuzzleWithState(visiblePuzzle.Puzzle, visiblePuzzle.State, content));
+                files.TryGetValue(puzzleView.ID, out ContentFile content);
+                puzzleView.Content = content;
+            }
+
+            if (ShowAnswers)
+            {
+                Dictionary<int, ContentFile> answers = await (from file in _context.ContentFiles
+                                                            where file.Event == Event && file.FileType == ContentFileType.Answer
+                                                            select file).ToDictionaryAsync(file => file.Puzzle.ID);
```

Comment:
This is likely a copy of the puzzle file over-query (but having this chunk of code does save us other overquerying once answers unlock)